pkg/actuators/machine/machine.go

```diff
@@ -0,0 +1,321 @@
+	if instance != nil {
+		klog.Infof("Skipped creating a VM that already exists.\n")
+		return nil
+	}
+
+	ignition, err := ms.getIgnition()
+	if err != nil {
+		return apierrors.CreateMachine("Error getting VM ignition: %v", err)
+	}
+
+	instance, err = ms.ovirtClient.CreateVmByMachine(ms.machine, ignition, ms.machineProviderSpec)
+	if err != nil {
+		return fmt.Errorf("Error creating Ovirt instance: %v", err)
+	}
+
+	// Wait till ready
+	// TODO: export to a regular function
+	// TODO: Why are we always setting error to nil?!
+	err = util.PollImmediate(RetryIntervalInstanceStatus, TimeoutInstanceCreate, func() (bool, error) {
+		instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+		if err != nil {
+			return false, nil
+		}
+		return instance.MustStatus() == ovirtsdk.VMSTATUS_DOWN, nil
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating oVirt VM: %v", err)
+	}
+
+	// Start the VM
+	err = ms.ovirtClient.StartVM(instance.MustId())
+	if err != nil {
+		return fmt.Errorf("Error running oVirt VM: %v", err)
+	}
+
+	// Wait till running
+	// TODO: export to a regular function
+	err = util.PollImmediate(RetryIntervalInstanceStatus, TimeoutInstanceCreate, func() (bool, error) {
+		instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+		if err != nil {
+			return false, nil
+		}
+		return instance.MustStatus() == ovirtsdk.VMSTATUS_UP, nil
+	})
+	if err != nil {
+		return fmt.Errorf("Error running oVirt VM: %v", err)
+	}
+	return nil
+}
+
+// exists returns true if machine exists.
+func (ms *machineScope) exists() (bool, error) {
+	instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+	if err != nil {
+		return false, err
+	}
+	return instance != nil, err
+}
+
+func (ms *machineScope) delete() error {
+	instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+	if err != nil {
+		return err
+	}
+	if instance == nil {
+		klog.Infof("Skipped deleting a VM that is already deleted.\n")
+		return nil
+	}
+
+	return ms.ovirtClient.DeleteVM(instance.MustId())
+}
+
+// returns the ignition from the userData secret
+func (ms *machineScope) getIgnition() ([]byte, error) {
+	if ms.machineProviderSpec == nil || ms.machineProviderSpec.UserDataSecret == nil {
+		return nil, nil
+	}
+	userDataSecret := &corev1.Secret{}
+	objKey := client.ObjectKey{
+		Namespace: ms.machine.Namespace,
+		Name:      ms.machineProviderSpec.UserDataSecret.Name,
+	}
+	if err := ms.client.Get(ms.Context, objKey, userDataSecret); err != nil {
+		return nil, err
+	}
+	userData, exists := userDataSecret.Data[userDataSecretKey]
+	if !exists {
+		return nil, fmt.Errorf("secret %s missing %s key", objKey, userDataSecretKey)
+	}
+	return userData, nil
+}
+
+func (ms *machineScope) patchMachine(ctx context.Context, condition ovirtconfigv1.OvirtMachineProviderCondition) error {
+	instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+	if err != nil {
+		return err
```

Comment:
Consider wrapping the error for context here.

Reply:
Done